File: Exercise1/Python/PFsample1.py
# ---
# jupyter:
#   jupytext:
#     text_representation:
#       extension: .py
#       format_name: light
#       format_version: '1.5'
#       jupytext_version: 1.7.1
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# +
'''
    PFsample1.py
    Sample Program for Particle Filter
    1. 2-D Random Walk
'''

# import modules
import random
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm

# fix seed
RANDOM_SEED = 0
random.seed(RANDOM_SEED)
np.random.seed(RANDOM_SEED)

# number of time points & particles
Nt = 100
Np = 10000

# variances of system noise (unknown)
tau_x = 0.5
tau_y = 0.2
v_cov = np.array([[tau_x,0],[0,tau_y]])

# variances of observation noise (known)
sigma_x = 0.1
sigma_y = 0.1
w_cov = np.array([[sigma_x,0],[0,sigma_y]])

# synthetic observation data
mean0 = np.array([0,0])
vt = np.random.multivariate_normal(mean0, v_cov, Nt)
wt = np.random.multivariate_normal(mean0, w_cov, Nt)
yt = np.cumsum(vt+wt, axis=0)

# +
# preparation for sequential estimation
x0 = np.zeros([Np,2])
xtp = np.zeros([Nt,Np,2])
xtf = np.zeros([Nt,Np,2])
llh = np.zeros([Nt,Np])
res = np.zeros(Np, dtype=int)

# sequential estimation using particle filter
for t in tqdm(range(Nt)):
    
    x = xtf[t-1] if t>0 else x0
    
    # one-step ahead prediction
    xtp[t] = x + np.random.multivariate_normal(mean0, v_cov, Np)
    
    # log-likelihood at this time point
    llh[t] = -np.log(2*np.pi) - 0.5*(
        np.log(sigma_x) + np.log(sigma_y)
        + (yt[t,0] - xtp[t,:,0])**2 / sigma_x
        + (yt[t,1] - xtp[t,:,1])**2 / sigma_y
    )
    
    # weight of each particle
    weight = np.exp(llh[t] - np.max(llh[t])) / np.sum(np.exp(llh[t] - np.max(llh[t])))
    
    # filtering based on the systematic residual resampling
    u = np.random.rand() / Np
    
    for p in range(Np):
        res[p] = np.floor((weight[p] - u) * Np) + 1
        u += res[p] / Np - weight[p]
    
    k = 0
    for i in range(Np):
        for j in range(res[i]):
            xtf[t,k] = xtp[t,i]
            k += 1

# log-likelihood of each model
llh_model = np.sum(llh, axis=1)
# -

# figures
plt.plot(yt[:,0], yt[:,1], label='true')
plt.plot(np.mean(xtf[:,:,0],axis=1), np.mean(xtf[:,:,1],axis=1), label='prediction')
plt.legend(loc='best')
plt.xlabel('x')
plt.ylabel('y')
plt.title('Particle Filter for 2-D Random Walk')
plt.show()

File: Exercise1/R/PFsample1.R
######################################
# PFsample1.R                        #
# Sample Program for Particle Filter #
#                                    #
# 1. 2-D random walk                 #
######################################

# parameters
Nt <- 100             # number of time points
Np <- 10000           # number of particles
tau2_T <- c(0.5,0.2)  # variances of system noise (unknown)
sigma2 <- c(0.1,0.1)  # variances of observation noise (known)

# synthetic observation data
yt <- rbind( cumsum( rnorm( Nt, mean=0, sd=sqrt(tau2_T[1]) ) ),
             cumsum( rnorm( Nt, mean=0, sd=sqrt(tau2_T[2]) ) ) ) +
      rbind( rnorm( Nt, mean=0, sd=sqrt(sigma2[1]) ),
             rnorm( Nt, mean=0, sd=sqrt(sigma2[2]) ) )

# sample particles
tau2 <- rbind( runif( Np, min=0, max=1 ), runif( Np, min=0, max=1 ) )

# preparation for sequential estimation
x0  <- c(0,0)                         # initial state
xtp <- array( NA, dim=c(2,Np,Nt) )    # predictive distribution
xtf <- array( NA, dim=c(2,Np,Nt) )    # filter distribution
llh <- matrix( NA, nrow=Np, ncol=Nt ) # log-likelihood at each time
res <- c( NA, length=Np )             # number of particles to be resampled

# sequential estimation using particle filter
for ( t in 1:Nt ) {
  if (t>1) { x <- xtf[ , ,t-1] } else { x <- x0 }

  # one-step ahead prediction
  xtp[ , ,t] <- x + rbind( rnorm( Np, mean=0, sd=sqrt( tau2[1, ] ) ),
                           rnorm( Np, mean=0, sd=sqrt( tau2[2, ] ) ) )

  # log-likelihood at this time point
  llh[ ,t] <- - log(2*pi) - 0.5 * (
                           log( sigma2[1] ) + log( sigma2[2] )
                         + ( yt[1,t] - xtp[1, ,t] )**2 / sigma2[1]
                         + ( yt[2,t] - xtp[2, ,t] )**2 / sigma2[2] )

  # weight of each particle
  weight <- exp( llh[ ,t] - max(llh[ ,t]) ) / sum( exp( llh[ ,t] - max(llh[ ,t]) ) )

  # filtering based on the systematic residual resampling
  u <- runif(1) / Np

  for ( p in 1:Np ) {
    res[p] <- floor( ( weight[p] - u ) * Np ) + 1
    u <- u + res[p] / Np - weight[p]
  }

  k <- 0
  for ( i in 1:Np ) {
  for ( j in 1:res[i] ) {
    if ( res[i]==0 ) break
    k <- k + 1
    xtf[ ,k,t] <- xtp[ ,i,t]
  }
  }
}

# log-likelihood of each model
llh_model <- apply( llh, 2, sum )

# figures
plot( yt[1, ], yt[2, ], xlab="x", ylab="y", type="l", col="black" )
lines( apply( xtf[1, , ], 2, mean ), apply( xtf[2, , ], 2, mean ),
       type="l", col="red" )
